refactor(@embark/core): move `Engine` into `embark-core`

PascalPrecht opened the pull request:
This also moves other modules and packages into `embark-core` which are
affected by the move of `Engine`.

It did some clean-up of some modules (e.g. `lib/utils/utils.js` had a lot of
stuff that was not used anywhere).

This is the first step to start typing our core inside out. For now we're just
moving code into their dedicated package, after that we can address making
use of TypeScript features as we see fit.

PascalPrecht commented on packages/core/core/src/engine.ts:
```diff
@@ -38,7 +41,7 @@ class Engine {
 
     if (this.interceptLogs || this.interceptLogs === undefined) {
-      utils.interceptLogs(console, this.logger);
+      interceptLogs(console, this.logger);
```
I've inlined `interceptLogs()` in this module because it's not used anywhere else within Embark.

PascalPrecht commented on packages/core/core/src/fs.ts:
```diff
@@ -0,0 +1,92 @@
+  return implementation(sourceFilePath, targetFilePath);
+}
+
+export function outputFileSync(...args) { return fs.outputFileSync(...args); }
```
AFAICS there's only two APIs that are specific to Embark (not coming from `fs-extra`). There good opportunity to get rid of most of what's in this module as most functions used are coming from `fs-extra`.

PascalPrecht commented on packages/core/core/src/index.js:
```diff
@@ -1,5 +1,12 @@
+export { EmbarkEmitter as Events } from './events';
+export { Plugins } from './plugins';
+export { TestLogger } from './utils/test_logger';
+export { TemplateGenerator } from './utils/template_generator';
```
There are a few symbols that should not be exported here, but are for the time being (didn't want to make this PR more complex as it just focusses on moving `Engine` and alike to `embark-core`:

- `TestLogger` should probably go into `embark-testing` at some point (if we really need it, otherwise we can drop it)
- `TemplateGenerator` is potentially something we can either move somewhere else, or have `Engine` expose an API that makes use of `TemplateGenerator` so that consumers can just import and use `Engine` (and don't have to worry about `TemplateGenerator`).

PascalPrecht commented on packages/core/core/src/utils/template_generator.js:
```diff
@@ -46,7 +51,7 @@ class TemplateGenerator {
     console.log(__('Installing template from ' + browse).green);
     console.log(__('Downloading template...').green);
-    fs.mkdirpSync(utils.dirname(tmpFilePath));
+    fs.mkdirpSync(path.dirname(tmpFilePath));
```
Turns out `utils.dirname()` didn't add any value over `path.dirname()` so I got rid off it as well

jrainville replied:
I think it was part of abstracting `path`, but the value it's not worth it

PascalPrecht commented on packages/core/core/src/utils/template_generator.js:
```diff
@@ -162,7 +167,7 @@ class TemplateGenerator {
 
   installTemplate(templatePath, name, installPackages, cb) {
-    utils.cd(templatePath);
+    shelljs.cd(templatePath);
```
This is the only place `utils.cd()` was used so I removed it and had `installTemplate()` make use of `shelljs.cd()` directly.

PascalPrecht commented on packages/core/core/src/utils/template_generator.js:
```diff
@@ -306,4 +311,9 @@ class TemplateGenerator {
+  decompress(filename, packageDirectory, opts).then((_files) => {
+    cb();
+  });
+}
```
This function used to live in `lib/utils/utils` but was nowhere else used, so I inlined it here.

PascalPrecht commented on packages/core/core/src/utils/utils.js:
```diff
@@ -0,0 +1,9 @@
+export function fileMatchesPattern(patterns, intendedPath) {
+  const globule = require('globule');
+  return globule.isMatch(patterns, intendedPath);
+}
```
This was originally in `lib/utils/utils` but most of what was in there could either be inlined (as commented above) or removed entirely because of lack of usage inside Embark. This leaves us with a utils file with only two functions that are actually shared inside this package.

PascalPrecht commented on packages/embark/src/cmd/cmd.js:
```diff
@@ -194,26 +194,6 @@ class Cmd {
-    //   });
-
-    // TODO: fix me, re-add above
-
```
This was a left over I assume so I removed it as well.

PascalPrecht commented on packages/embark/src/cmd/cmd_controller.js:
```diff
@@ -1,12 +1,14 @@
-const fs = require('../lib/core/fs.js');
 const cloneDeep = require('clone-deep');
 
+setUpEnv(joinPath(__dirname, '../../'));
```
`setUpEnv()` is the function that ensures `DAPP_PATH` and `NODE_PATH` are set correctly. We used to call it inside `env.js`. However, `env.js` used to be loaded in `lib/core/config` but since we moved that one to `embark-core` we need to ensure `setUpEnv()` is still called at a place when Embark is intialized (to ensure env variables are set correctly). That's why this is in `cmd_controller` now

PascalPrecht commented on packages/embark/src/cmd/dashboard/dashboard.js:
```diff
@@ -1,5 +1,12 @@
+    return windowSize.get();
+  }
+  return {width: 240, height: 75};
+}
```
`getWindowSize()` was only used here, so I've inlined it.

PascalPrecht commented on packages/embark/src/test/modules/solidity/remapImports.js:
```diff
@@ -2,7 +2,6 @@
 const { dappPath, File, Types, prepareForCompilation } = require('embark-utils');
 const path = require("path");
 const {expect} = require("chai");
-const fs = require("../../../lib/core/fs");
```
There was no usage of `fs` in the entire module as it relies on `fsNode` (one line below), so this could be safely removed.

michaelsbradleyjr commented on dapps/templates/demo/config/communication.js:
```diff
@@ -1,7 +1,7 @@
   // default applies to all environments
   default: {
-    enabled: true,
+    enabled: false,
```
Does this mean whisper will be disabled by default in embark_demo?

PascalPrecht replied:
This was for testing purposes and is a left over. Will undo this change again. Good catch!

michaelsbradleyjr commented on packages/core/core/.eslintignore:
```diff
@@ -0,0 +1,2 @@
+src/**/*.ts
```
Was eslint actually checking the `.ts` files. If so, that's very interesting; I though it just ignored anything that's not a `.js` file.

PascalPrecht replied:
This is also a leftover. Chances are, I confused tsc, tslint, eslint + babel parsing errors as I went through this. Will remove this.

michaelsbradleyjr commented on packages/core/core/src/fs.ts:
```diff
@@ -0,0 +1,93 @@
+
+export function statSync(args: any, args2?: any) { return fs.statSync(args, args2); }
+
+export function readJSONSync(args: any, args2?: any) {
```
fs-extra provides a [`readJSONSync`](https://github.com/jprichardson/node-fs-extra/blob/master/lib/json/index.js#L14) function.

I don't recall if there was a special reason we implement it here instead of using fs-extra's implementation. It may be that at the time we were using an older version of fs-extra that didn't have that function, not really sure (cc: @iurimatias, @jrainville — any ideas?).

If there isn't a good reason for having our own, I think we should forward calls to fs-extra's function instead. Also, in the case the caller wants the behavior where exceptions are caught, an error message is logged and the process exits, I think that should be the responsibility of the caller or provided by a util function — it seems like a bad idea for that behavior to be in the `fs` function directly (though I may have been the one that added it back in the day, I sort of recall doing that).

PascalPrecht replied:
I agree. This module should be revisited once the move to `embark-core` was successful.

jrainville replied:
No idea. I agree that we use fs-extra

michaelsbradleyjr commented on packages/core/core/src/fs.ts:
```diff
@@ -0,0 +1,93 @@
+export function createWriteStream(args: any, args2?: any) { return fs.createWriteStream(args, args2); }
+
+export function copyPreserve(sourceFilePath, targetFilePath) {
+  const implementation = (sourceFilePath, targetFilePath) => {
```
I wrote this, but I don't remember why I define the `implementation` function and then call it instead of just having the code inside `implementation` be the body of `copyPreserve`. It may be that initially I was going to use recursion instead of a `while` loop. In any case, just noting that it's something we can clean up if we wish.

michaelsbradleyjr commented on packages/core/core/src/utils/template_generator.js:
```diff
@@ -260,7 +265,7 @@ class TemplateGenerator {
       if (templateAndBranch.length === 1) {
         fallback = true;
-        embarkVersion = semver(require('../../../package.json').version);
+        embarkVersion = semver(require('../../../../../package.json').version);
```
It is possible for the core package's version to lag behind `packages/embark`, i.e. if in a minor/patch bump the core package didn't change but `packages/embark` did. So I suggest this instead:

```suggestion
        embarkVersion = semver(require(embarkPath('package.json')).version);
```

PascalPrecht replied:
Thanks, will do!

PascalPrecht commented on packages/core/core/package.json:
```diff
@@ -35,28 +35,43 @@
     "ci": "npm run qa",
     "clean": "npm run reset",
     "lint": "eslint process.js src/",
```
👍

PascalPrecht replied:
I updated this locally (still have to push), but I'll disable `lint:ts` for the time being as it raises several pages of lint errors at the moment that aren't scope of this PR. 

Will address this after the move into `embark-core` is done

PascalPrecht commented on packages/core/core/package.json:
```diff
@@ -35,28 +35,43 @@
   "devDependencies": {
+    "@babel/core": "7.6.4",
+    "babel-jest": "24.9.0",
+    "clone-deep": "4.0.0",
```
I thought it was needed to make the collective babel config work, which uses `clone-deep`. Will remove this!

PascalPrecht commented on packages/core/core/src/engine.ts:
```diff
@@ -21,27 +71,34 @@ class Engine {
-    this.ipcRole = options.ipcRole || 'client';
+    this.ipcRole = options.ipcRole;
+    this.config = new Config({env: this.env, logger: this.logger, events: this.events, context: this.context, webServerConfig: this.webServerConfig, version: this.version});
+    this.isDev = this.config && this.config.blockchainConfig && (this.config.blockchainConfig.isDev || this.config.blockchainConfig.default);
```
Yes. All for that!

jrainville replied:
Oh wow that's very cool. TIL

PascalPrecht commented:
Addressed the comments and made CI green locally.

Notice that I've turned of tslint for `embark-core` for now. Will address that in a follow up PR.

PascalPrecht commented:
I've rebased this on top of #1986 to make CI go green. Hence the additional commits. We can either decide to merge this one and close #1986, or land #1986 first and have this one rebased on top of `master` again.

PascalPrecht commented:
Some `qa` tests were still failing that weren't failing/run locally. Updated the PR accordingly once more.

PascalPrecht commented:
`embark run` command seems to be hanging now. Surprised `yarn ci` didn't catch this locally. Looking into this.

PascalPrecht commented:
Okay, the reason it was hanging was because the `processManager` wasn't initialised, making Embark waiting for a response of an event that practically didn't exist.

I've created an issue here, so we don't forget to introduce proper error handling for cases where events that don't exist are requested: https://github.com/embark-framework/embark/issues/2025

PascalPrecht commented:
This is now ready for merge

iurimatias commented:
@PascalPrecht could you rebase plz.

PascalPrecht commented:
Working on it.

PascalPrecht commented:
Rebased.

michaelsbradleyjr commented on packages/core/code-runner/package.json:
```diff
@@ -52,6 +52,9 @@
     "embarkjs": "^5.0.0-alpha.1",
+    "embark-core": "^5.0.0-alpha.1",
+    "embark-logger": "^5.0.0-alpha.1",
+    "embarkjs": "^5.0.0-alpha.1",
```
`"embarkjs"` is already spec'd on L54.

PascalPrecht replied:
Good catch! Thanks

michaelsbradleyjr commented on packages/core/core/package.json:
```diff
@@ -34,29 +34,44 @@
-    "qa": "npm-run-all lint _build",
+    "lint": "npm-run-all lint:*",
+    "lint:js": "eslint process.js src/",
+    "// lint:ts": "tslint -c tslint.json \"src/**/*.ts\"",
```
I'm assuming that `lint:ts` is kept disabled for now because various linter errors will be fixed in later PRs.

Even so, if in `packages/core/core` you run `tslint -c tslint.json \"src/**/*.ts\" --fix` it will cleanup quite a few automatically:

```
$ tslint -c tslint.json "src/**/*.ts" --fix
Fixed 116 error(s) in src/config.ts
Fixed 9 error(s) in src/configDefaults.ts
Fixed 16 error(s) in src/engine.ts
Fixed 15 error(s) in src/events.ts
Fixed 21 error(s) in src/plugin.ts
Fixed 27 error(s) in src/plugins.ts
Fixed 12 error(s) in src/services_monitor.ts
```

It might be worth including those fixes in this PR, but I'll leave it up to you, this isn't a change request.

PascalPrecht replied:
Thanks, adding this!

PascalPrecht replied:
Actually, it doesn't fix a lot of them, so I'll keep it disabled for now. Will get those changes in in another PR that addresses type satefy as well

PascalPrecht replied:
In fact, running this command doesn't give me the same output as yours

michaelsbradleyjr replied:
Okay... but it should give identical output. I realize now the command example may have been unclear; more precisely:

```
$ cd packages/core/core
$ npx tslint -c tslint.json "src/**/*.ts" --fix
Fixed 116 error(s) in src/config.ts
Fixed 9 error(s) in src/configDefaults.ts
Fixed 16 error(s) in src/engine.ts
Fixed 15 error(s) in src/events.ts
Fixed 21 error(s) in src/plugin.ts
Fixed 27 error(s) in src/plugins.ts
Fixed 12 error(s) in src/services_monitor.ts
```

The output is what I got just now after re-fetching this PR branch.

PascalPrecht replied:
Thanks @michaelsbradleyjr for some reason, this didn't work before (even with `npx`), but now it did. Updating..

PascalPrecht commented:
Updated it once more
